### ISSUE TITLE
Load Custom Datasets for evaluation 

### ISSUE THREAD

**Issue by jas-nat** [facebookresearch/detectron2#552]
I am trying to evaluate my data using different .py file for training and testing. Although I have registered the metadata for validation in the training file, it isn't recognized in the testing file. Do we need to register the custom datasets every time I train and evaluate custom datasets?

**Comment by ppwwyyxx** [comment]
The code that registers the dataset needs to be executed in each process in order to take effect in that process.

**Comment by jas-nat** [comment]
Okay got it. It means if I have trained the data and save the model, I just need to register the dataset again, without training then I can evaluate directly. Am I correct?

**Comment by ppwwyyxx** [comment]
If it's a new process, then you need to register it again.

**Comment by jas-nat** [comment]
## Cannot load the model
1. what changes you made
I used the custom datasets and try to train. My datasets have 9 labels, but what I want to detect is only 1 label. I specify it in the function of registration by using the conditional statement. The function is like
```
annotation = "../annotations/train.json"

def get_KITTI_dicts(img_path):
	json_path = os.path.join(img_path, annotation)
	with open(json_path) as f:
	 	for i, line in enumerate(f):
	 		images = json.loads(line) # run through each line

	dataset_dicts = []
	
	for i, image in enumerate(images['images']):
		record = {}
		print("registering image")
		
		filename = os.path.join(img_path, image["file_name"])
		_id = image["id"]

		#if the file doesn't exist
		try:
			height, width = cv2.imread(filename).shape[:2]
		except AttributeError:
			print("File doesn't exist")
			continue

		record["file_name"] = filename
		record["height"] = height
		record["width"] = width
		record["image_id"] = _id

		objs = [] #many instances in 1 record
						
		for anno in images['annotations']:
			anno_id = anno["image_id"]
			category_id = anno["category_id"]
			#check if the image id from image data same with annotation and include only person and person sitting
			#although in the JSON file it is number 8, the annotation start from 1
			if anno_id == _id and (category_id == 8 or category_id == 3): 
				area = anno["area"]	
				instance_id = anno["id"]
				print("Iter {2} Instance {1} In image {0}".format(filename, instance_id,i))	
				px = anno["bbox"][0]
				py = anno["bbox"][1]
				p_width = anno["bbox"][2]
				p_height = anno["bbox"][3]
			
				obj = {"bbox": [px,py,p_width,p_height],
						"bbox_mode": BoxMode.XYWH_ABS, #it's not XYXY but XYWH
						"area": area,
						"segmentation":[],
						#"category_name": category,
						"category_id": category_id,
						"iscrowd": 0}
				objs.append(obj)

		record["annotations"] = objs
		dataset_dicts.append(record)
		
	return dataset_dicts
```
I register the dataset as following:

```
for d in ["train","val"]:
    DatasetCatalog.register("KITTI_Object_" + d, lambda d=d: get_KITTI_dicts("../datasets/coco/" + d))
    MetadataCatalog.get("KITTI_Object_" + d).set(thing_classes=["Cyclist", "DontCare", "Misc","Person_sitting", "Tram", "Truck", "Van", "car", "person"])
KITTI_metadata = MetadataCatalog.get("KITTI_Object_train")
```

My configuration file is :
```
cfg = get_cfg()
# cfg.merge_from_file("./configs/COCO-Detection/retinanet_R_50_FPN_1x.yaml") #try to use Retina Net 50 1x
cfg.DATASETS.TRAIN = ("KITTI_Object_train",)
cfg.DATASETS.TEST = ()   # no metrics implemented for this dataset
cfg.DATALOADER.NUM_WORKERS = 2
cfg.MODEL.WEIGHTS = "detectron2://ImageNetPretrained/MSRA/R-50.pkl"
cfg.SOLVER.IMS_PER_BATCH = 2 #each GPU will see 2 image per batch
cfg.SOLVER.BASE_LR = 0.001 #learning rate
cfg.SOLVER.MAX_ITER = 10000    # 300 iterations seems good enough, but you can certainly train longer


cfg.MODEL.ROI_HEADS.BATCH_SIZE_PER_IMAGE = 128   # faster, and good enough for this toy dataset
cfg.MODEL.RETINANET.NUM_CLASSES = 9 
```
My testing configuration =
```
os.makedirs(cfg.OUTPUT_DIR, exist_ok=True)
trainer = DefaultTrainer(cfg) 
trainer.resume_or_load(resume=False)
trainer.train() #start training

cfg.MODEL.WEIGHTS = os.path.join(cfg.OUTPUT_DIR, "model_final.pth") #original
cfg.MODEL.RETINANET.SCORE_THRESH_TEST = 0.5 #percentage above 50%
cfg.MODEL.RETINANET.NMS_THRESH_TEST = 0.5
cfg.DATASETS.TEST = ("KITTI_Object_val", )

predictor = DefaultPredictor(cfg) #evaluation
```
It could show the prediction quite well by drawing the instance predictions. The images also can be saved.
Then, I tried to evaluate using COCO evaluator based on the code in Google collab

2. what exact command you run:
Since I asked at #277 that the model is saved in "model_final.pth", I rerun the whole code except I comment these codes:
```os.makedirs(cfg.OUTPUT_DIR, exist_ok=True)
trainer = DefaultTrainer(cfg) 
trainer.resume_or_load(resume=False)
trainer.train() 
```
I thought I have loaded the datafile in `cfg.MODEL.WEIGHTS=os.path.join(cfg.OUTPUT_DIR, "model_final.pth"`. In other words, I don't want to train the model again when evaluating

3. what you observed (including the full logs):
I always get 0 AP and AR when evaluating.
![failed](https://user-images.githubusercontent.com/52592216/71255002-08d97600-2370-11ea-9e27-788413f12e35.png)

When I train and test at the same time, I can get the result as follows:
![AP 50 1x 10000](https://user-images.githubusercontent.com/52592216/71254507-8b613600-236e-11ea-9225-4267f50c92e3.png)


## Expected behavior
1. Does it mean I have to always train and evaluate the dataset at the same time? 
2. Do I load the data incorrectly?
3. How are the AP and AR computed? Is it based on all labels that exist in the ground truth labels?
4. About my environment, I have tried to reinstall my CUDA, pytorch, and Detectron2 again. However, my Detectron 2 CUDA Compiler is still not detected. Does that matter? Although it is not available, I could run the balloon example codes locally. 
## Environment

------------------------  --------------------------------------------------
sys.platform              linux
Python                    3.6.8 (default, Oct  7 2019, 12:59:55) [GCC 8.3.0]
Numpy                     1.17.4
Detectron2 Compiler       GCC 7.4
Detectron2 CUDA Compiler  not available
DETECTRON2_ENV_MODULE     <not set>
PyTorch                   1.3.1
PyTorch Debug Build       False
torchvision               0.4.2
CUDA available            True
GPU 0                     GeForce RTX 2070 SUPER
CUDA_HOME                 /usr
NVCC                      Cuda compilation tools, release 9.1, V9.1.85
Pillow                    6.2.1
cv2                       3.2.0
------------------------  --------------------------------------------------
PyTorch built with:
  - GCC 7.3
  - Intel(R) Math Kernel Library Version 2019.0.4 Product Build 20190411 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v0.20.5 (Git Hash 0125f28c61c1f822fd48570b4c1066f96fcb9b2e)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - NNPACK is enabled
  - CUDA Runtime 10.1
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_37,code=compute_37
  - CuDNN 7.6.3
  - Magma 2.5.1
  - Build settings: BLAS=MKL, BUILD_NAMEDTENSOR=OFF, BUILD_TYPE=Release, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -fopenmp -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -O2 -fPIC -Wno-narrowing -Wall -Wextra -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Wno-stringop-overflow, DISABLE_NUMA=1, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, USE_CUDA=True, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, USE_STATIC_DISPATCH=OFF,

**Comment by ppwwyyxx** [comment]
* To load a model, you need https://detectron2.readthedocs.io/tutorials/models.html
* I think it will only evaluate the classes that you registered into the metadata.
If you run a retinanet, then the missing cuda compiler does not matter. If not, the model will be very slow.

**Comment by jas-nat** [comment]
> * To load a model, you need https://detectron2.readthedocs.io/tutorials/models.html
> * I think it will only evaluate the classes that you registered into the metadata.
> * If you run a retinanet, then the missing cuda compiler does not matter. If not, the model will be very slow.

- I see that to load the models I haven't implemented these:
```
from detectron2.modeling import build_model
model = build_model(cfg)  # returns a torch.nn.Module
```
What is the difference with the codes to load the model from the model zoo? Because I assume by running these two line codes, I have loaded the model. 
`cfg.merge_from_file("./configs/COCO-Detection/retinanet_R_50_FPN_1x.yaml")`
`cfg.MODEL.WEIGHTS = "detectron2://ImageNetPretrained/MSRA/R-50.pkl"`

**Comment by ppwwyyxx** [comment]
1. Setting the config let trainer / predictor know what to load. `build_model` does not load weights.
2. register the desired labels in the metadata.

**Comment by jas-nat** [comment]
> 2. register the desired labels in the metadata.

I modified my code by put `"category_id" : 0` in the obj annotation. I also set `cfg.MODEL.RETINANET.NUM_CLASSES = 1 ` and changed the metadata by 
```
for d in ["train","val"]:
    DatasetCatalog.register("KITTI_Object_" + d, lambda d=d: get_KITTI_dicts("../datasets/coco/" + d))
    MetadataCatalog.get("KITTI_Object_" + d).set(thing_classes=["person"])
```
However, my result was still the same as this


What did I miss?

**Comment by ppwwyyxx** [comment]
You're making progress: now it only evaluates one class.

https://detectron2.readthedocs.io/tutorials/models.html says to load a model you need `DetectionCheckpointer(model).load(file_path)`. That does not seem like what you did.

**Comment by jas-nat** [comment]
If I use DefaultTrainer and DefaultPredictor, are the returned values the same as build_model? Because I still get the same results as before. I used this code

```
print("Start Training in 5 seconds")
time.sleep(5)

# see config reference from https://detectron2.readthedocs.io/modules/config.html#config-references
cfg = get_cfg()
cfg.merge_from_file("./configs/COCO-Detection/retinanet_R_50_FPN_1x.yaml") #try to use Retina Net 50 1x
#cfg.merge_from_file("./configs/COCO-Detection/retinanet_R_50_FPN_3x.yaml") #try to use Retina Net 50 1x
# cfg.merge_from_file("./configs/COCO-Detection/retinanet_R_101_FPN_3x.yaml") #retinanet R101 3x
print("Begin DATASETS.TRAIN in 3 seconds")
time.sleep(3)
cfg.DATASETS.TRAIN = ("KITTI_Object_train",)

print("cfg.DATASETS.TEST")
cfg.DATASETS.TEST = ()   # no metrics implemented for this dataset


print("cfg.DATASETS.NUM_WORKERS")
# Number of data loading threads
cfg.DATALOADER.NUM_WORKERS = 2

# #change the model id
# #following the repository format in ...yaml file, weight key
cfg.MODEL.WEIGHTS = "detectron2://ImageNetPretrained/MSRA/R-50.pkl"
# cfg.MODEL.WEIGHTS = "detectron2://ImageNetPretrained/MSRA/R-101.pkl"

cfg.SOLVER.IMS_PER_BATCH = 2 #each GPU will see 2 image per batch
cfg.SOLVER.BASE_LR = 0.001 #learning rate
cfg.SOLVER.MAX_ITER = 2000    # 300 iterations seems good enough, but you can certainly train longer

# RoI minibatch size *per image* (number of regions of interest [ROIs])
# Total number of RoIs per training minibatch =
#   ROI_HEADS.BATCH_SIZE_PER_IMAGE * SOLVER.IMS_PER_BATCH
# E.g., a common configuration is: 512 * 16 = 8192
cfg.MODEL.ROI_HEADS.BATCH_SIZE_PER_IMAGE = 128   # faster, and good enough for this toy dataset
# cfg.MODEL.ROI_HEADS.NUM_CLASSES = 2 #because there are 9 classifications 
cfg.MODEL.RETINANET.NUM_CLASSES = 1 #a person only 

#load the checkpoints and model
from detectron2.modeling import build_model
from detectron2.checkpoint import DetectionCheckpointer
model = build_model(cfg) #build the model

print("cfg finished")

#directory where output file is written
#./output
os.makedirs(cfg.OUTPUT_DIR, exist_ok=True)
trainer = DefaultTrainer(cfg) 
trainer.resume_or_load(resume=False)
trainer.train() #start training

#losses saved still not running
# from detectron2.utils.events import EventStorage
# with EventStorage() as storage:
# 	losses = model(dataset_dicts)

print("finished train")
time.sleep(5)
'''
TESTING

go to KITTI test_ped.py
'''
cfg.MODEL.WEIGHTS = os.path.join(cfg.OUTPUT_DIR, "model_final.pth") #original
# cfg.MODEL.WEIGHTS = os.path.join(cfg.OUTPUT_DIR, "KITTI_ped_resnet50.pth") #rename the model after training
cfg.MODEL.RETINANET.SCORE_THRESH_TEST = 0.5 #percentage above 50%
cfg.MODEL.RETINANET.NMS_THRESH_TEST = 0.5
cfg.DATASETS.TEST = ("KITTI_Object_val", )
DetectionCheckpointer(model).load("./output/model_final.pth")

predictor = DefaultPredictor(cfg) #evaluation


# Then, we randomly select several samples to visualize the prediction results.

dataset_dicts = get_KITTI_dicts("../datasets/coco/val")
model = build_model(cfg) #build the model
```

**Comment by ppwwyyxx** [comment]
I couldn't understand what you observed and what unexpected issue you're seeing. It would be
clearer if you could include details about the problem following the [issue template](https://github.com/facebookresearch/detectron2/issues/new?template=unexpected-problems-bugs.md).
At least telling from the code, you build a new model in the end without loading weights to it so the model cannot do anything.

**Comment by jas-nat** [comment]
## Evaluations results:

1. what changes you made (`git diff`) or what code you wrote
```

```
2. what exact command you run:

3. what you observed (including the full logs):
The result from [previous reply](https://github.com/facebookresearch/detectron2/issues/552#issuecomment-568517677)

## Expected behavior:
> You're making progress: now it only evaluates one class.
> 
> https://detectron2.readthedocs.io/tutorials/models.html says to load a model you need `DetectionCheckpointer(model).load(file_path)`. That does not seem like what you did.
I want to run an evaluation on my algorithm. Beforehand, the problem occured because I registered all objects to the dataset, but I actually only evaluated 1 object, so the evaluation did not work well.  Based on the reply, I have to load the model using the checkpointer.

I have tried, but I am not sure did I load the model at the correct position or not, since the result remains the same.

**Comment by ppwwyyxx** [comment]
The use of evaluation seems correct to me. You could be getting 0 score simply because the model is not trained well.

> the pictures are not saved.

then you should check your usage of opencv `imwrite`.

In general if you don't understand what the code is doing it's best to start with a working example (like balloon) and make changes only one at a time until it stops working.

**Comment by tamyxgopenx** [comment]
@ppwwyyxx  How about -100.0 value, what is this score support to be?
